### PR TITLE
Add support for json, yaml and xml

### PR DESCRIPTION
**Pull request by MatteoGheza**
This commit need testing

### CODE REVIEW

File: Config.php [hunk 1]
```diff
@@ -67,12 +67,48 @@ public function __construct(\PDO $dbh, $config_source = NULL, $config_type = '',
...
+                if (empty($config_source)) die('PHPAuth: config type is FILE, but no source file declared!'); //@todo: \Exception
+
+                // replace beginner '$' in filepath to application root directory
+                $source = preg_replace('/^\$/', getcwd(), $config_source);
+
+                // check json-config is readable
+                if (!is_readable($source)) die("PHPAuth: config type is FILE, declared as {$source}, but file not readable or not exist"); //@todo: \Exception
+
+                // load configuration
+                $this->config = json_decode(file_get_content($source));
```
Comment (turbopixel) [inline Config.php:80]: It's useful to check for JSON errors, I think. 
https://www.php.net/manual/en/function.json-last-error.php

File: Config.php [hunk 2]
```diff
@@ -67,12 +67,48 @@ public function __construct(\PDO $dbh, $config_source = NULL, $config_type = '',
...
+                if (empty($config_source)) die('PHPAuth: config type is FILE, but no source file declared!'); //@todo: \Exception
+
+                // replace beginner '$' in filepath to application root directory
+                $source = preg_replace('/^\$/', getcwd(), $config_source);
+
+                // check xml-config is readable
+                if (!is_readable($source)) die("PHPAuth: config type is FILE, declared as {$source}, but file not readable or not exist"); //@todo: \Exception
+
+                // load configuration
+                $this->config = simplexml_load_string(file_get_content($source));
```
Comment (turbopixel) [inline Config.php:110]: Validate for bad xml https://www.php.net/manual/en/domdocument.validate.php

**Comment by MatteoGheza** [comment]
Can you check the last commit?

File: Config.php [hunk 3]
```diff
@@ -106,8 +117,10 @@ public function __construct(\PDO $dbh, $config_source = NULL, $config_type = '',
                 // check xml-config is readable
                 if (!is_readable($source)) die("PHPAuth: config type is FILE, declared as {$source}, but file not readable or not exist"); //@todo: \Exception
 
-                // load configuration
-                $this->config = simplexml_load_string(file_get_content($source));
+                if (!XMLReader::open($source)->isValid()) die("PHPAuth: config type XML is corrupted"); //@todo: \Exception
```
Comment (turbopixel) [inline Config.php:120]: This doesn't work by me. I always get false back from `isValid()`. I found nothing in the documentation, but on Stackoverflow (https://stackoverflow.com/a/4554256/2310353)

This code works fine:
```
libxml_use_internal_errors(true);

$xmlstr = simplexml_load_file("test.xml");
if (!$xmlstr) {
  die("PHPAuth: config type XML is corrupted");
  // $errors = libxml_get_errors(); // <- returns xml erros
}
```

**So I think the following code is right:**

```
libxml_use_internal_errors(true); // don't display errors
$config = simplexml_load_file($source);

if(!$config){
   die("PHPAuth: config type XML is corrupted");
}

$this-config = $config;
```

**Comment by MatteoGheza** [comment]
ok

**Comment by turbopixel** [comment]
Looks fine!

**Comment by MatteoGheza** [comment]
Ok, can you approve the PR?

**Comment by MatteoGheza** [comment]
Thanks!

**Comment by MatteoGheza** [comment]
@KarelWintersky is this ok? Can you merge into master?

**Comment by KarelWintersky** [comment]
@turbopixel , we must drop support of PHP5.6 

See https://www.php.net/supported-versions.php

**Comment by louis123562** [comment]
7.4 should be minimum.

**Comment by KarelWintersky** [comment]
> 7.4 should be minimum.

Nope. 7.2 minimum, 7.3+ recommended.

**Comment by KarelWintersky** [comment]
@MatteoGheza , yaml_parse() is PECL function. It not included to all configurations.

You must check, is exists function `yaml_parse`, and if NOT - script will die with error message.

**Comment by Conver** [comment]
Status?

**Pull request closed without merging**